Doc `RawTableInner::find_insert_slot` and also make it unsafe

JustForFun88 opened the pull request:
Made this function unsafe, as it is possible to use the index returned by this function unsafely. Here are two functions that confirm the documentation I made:
```rust
#[test]
fn test_infinite_loop() {
	use ::alloc::vec::Vec;
	let mut table: RawTable<u16> = RawTable::with_capacity(20);

	assert_eq!(table.capacity(), 28);
	assert_eq!(table.buckets(), 32);
	for i in 0..table.buckets() {
		unsafe {
			let bucket_index = table.table.find_insert_slot(i as u64);

			table.table.set_ctrl_h2(bucket_index, i as u64);
			table.table.items += 1;

			table.bucket(bucket_index).write(i as u16);
		}
	}

	let vec = unsafe { table.iter().map(|x| x.read()).collect::<Vec<_>>() };

	assert_eq!(vec, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]);

	// this never return
	let _bucket_index = unsafe { table.table.find_insert_slot(32) };
}

#[test]
fn test_undefined_behavior() {
	use ::alloc::vec::Vec;

	let mut table: RawTable<u16> = RawTable::with_capacity(3);

	assert_eq!(table.capacity(), 3);
	assert_eq!(table.buckets(), 4);
	for i in 0..3 {
		table.insert(i, i as u16, |i| *i as u64);
	}

	assert_eq!(table.table.items, 3);
	assert_eq!(table.table.growth_left, 0);

	let bucket_index = unsafe { table.table.find_insert_slot(3_u64) };

	assert_eq!(bucket_index, 3);

	unsafe {
		table.table.set_ctrl_h2(bucket_index, 3_u64);
		table.table.items += 1;
		table.bucket(bucket_index).write(3_u16);
	}

	let vec = unsafe { table.iter().map(|x| x.read()).collect::<Vec<_>>() };

	assert_eq!(vec, [0, 1, 2, 3]);
	let bucket_index = unsafe { table.table.find_insert_slot(4_u64) };
	assert_eq!(bucket_index, 4);
}
```

Amanieu commented:
This function should not be unsafe. As you said in the doc, it is always safe to call. `unsafe` should only be used when a function can cause UB with certain arguments.

JustForFun88 commented:
> This function should not be unsafe. As you said in the doc, it is always safe to call. `unsafe` should only be used when a function can cause UB with certain arguments.

I removed unsafe and changed "Safety" to "Note"

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 1605442376c4223c8d9f3e967dae0544be9088f7 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/hashbrown) for this repository.

<!-- @bors r=Amanieu 1605442376c4223c8d9f3e967dae0544be9088f7 -->
<!-- homu: {"type":"Approved","sha":"1605442376c4223c8d9f3e967dae0544be9088f7","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/hashbrown"} -->

bors commented:
:hourglass: Testing commit 1605442376c4223c8d9f3e967dae0544be9088f7 with merge d40ffa1ad39c424bd24ad593bc3b488ccaebd06a...
<!-- homu: {"type":"BuildStarted","head_sha":"1605442376c4223c8d9f3e967dae0544be9088f7","merge_sha":"d40ffa1ad39c424bd24ad593bc3b488ccaebd06a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/hashbrown/actions/runs/4252731403/jobs/7396876765)
Approved by: Amanieu
Pushing d40ffa1ad39c424bd24ad593bc3b488ccaebd06a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/hashbrown/actions/runs/4252731403/jobs/7396876765"},"merge_sha":"d40ffa1ad39c424bd24ad593bc3b488ccaebd06a"} -->
